File: spark16/src/main/java/javasql/JDBCDataSource.java
package javasql;

import org.apache.commons.configuration.ConfigurationException;
import org.apache.commons.configuration.PropertiesConfiguration;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.DataFrame;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by 小省 on 2016/7/5.
 */
public class JDBCDataSource {
	public static void main(String[] args) throws ConfigurationException {
		SparkConf conf =new SparkConf().setAppName("JDBCDataSource").setMaster("local")
				;
		JavaSparkContext sc =new JavaSparkContext(conf);
		SQLContext sqlContext=new SQLContext(sc);
		
		PropertiesConfiguration pro =new PropertiesConfiguration("mysql.properties");
		
		String url=pro.getString("url");
				
		Map<String,String> option= new HashMap<>();
		option.put("url",url);
		option.put("dbtable","score");
		option.put("user","root");
		option.put("password","123456");

	DataFrame studentScoresDF=sqlContext.read().format("jdbc").options(option).load();

		option.put("dbtable","student");
		DataFrame studentInfosDF = sqlContext.read().format("jdbc")
				.options(option).load();

		/*
		 读取字段顺序，跟mysql中表的字段顺序是一致的
	    */
		JavaPairRDD<String, Tuple2<Integer, Integer>> studentsRDD = studentInfosDF.javaRDD().mapToPair(x -> new Tuple2<String, Integer>(
				x.getString(0), Integer.valueOf(String.valueOf(x.get(1)))))
				.join(
						studentScoresDF.javaRDD().mapToPair(x -> new Tuple2<String, Integer>(
								x.getString(0), Integer.valueOf(x.getInt(1))
						))
				);
		JavaRDD<Row> studentsRowRDD = studentsRDD.map(x -> RowFactory.create(x._1, x._2._1, x._2._2))

				.filter(x -> Integer.valueOf(String.valueOf(x.get(2))) > 87);

//schme
		List<StructField> fileds=new ArrayList<>();
		fileds.add(DataTypes.createStructField("name", DataTypes.StringType,true));
		fileds.add(DataTypes.createStructField("age", DataTypes.IntegerType,true));
		fileds.add(DataTypes.createStructField("score",DataTypes.IntegerType,true));

		StructType schema = DataTypes.createStructType(fileds);
		DataFrame studentsDF = sqlContext.createDataFrame(studentsRowRDD, schema);


		Row[] row=studentsDF.collect();
		
		for (Row row1 : row) {
			System.out.println(row1);

		}
//		//foreachPatition
//		studentsDF.javaRDD().foreach(new VoidFunction<Row>() {
//			@Override
//			public void call(Row row) throws Exception {
//				String sql = "insert into good_student values("
//						+ "'" + row.getString(0) + "',"
//						+ Integer.valueOf(String.valueOf(row.get(1))) + ","
//						+ Integer.valueOf(String.valueOf(row.get(2))) + ")" ;
//				Class.forName("com.mysql.jdbc.Driver");
//				Connection conn = null;
//				Statement stat = null;
//				try {
//					conn = DriverManager.getConnection("jdbc:mysql://centos1:3306/test","root","123456");
//					stat = conn.createStatement();
//					stat.executeUpdate(sql);
//				} catch (Exception e) {
//					e.printStackTrace();
//				}finally{
//					if(stat != null){
//						stat.close();
//					}
//					if(conn != null){
//						conn.close();
//					}
//				}
//
//			}
//		});

		//一个partition 一提交
//	studentsDF.javaRDD().foreachPartition(new VoidFunction<Iterator<Row>>() {
//		@Override
//		public void call(Iterator<Row> rowIterator) throws Exception {
//
//		}
//	});

	}
}

File: spark16/src/main/java/scasql/JDBCDataSource.scala
package scasql

import java.io.File
import java.nio.file.{Files, Path}
import java.sql.DriverManager
import java.util

import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
import org.apache.spark.sql.{Row, SQLContext}
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.sql.hive.HiveContext;

/**
  * Created by 小省 on 2016/7/6.
  */
object JDBCDataSource {
  def main(args: Array[String]) {
    val conf = new SparkConf().setAppName("JDBCDataSource").setMaster("local")
    val sc = new SparkContext(conf)
    val sqlContext = new SQLContext(sc);

    val options = new util.HashMap[String, String]()
    options.put("url","jdbc:mysql://centos1:3306/test")
    options.put("user", "root")
    options.put("password", "123456")
    options.put("dbtable", "student")

    val studentInfo = sqlContext.read.options(options).format("jdbc").load()
    options.put("dbtable", "score")

    val studentScore = sqlContext.read.options(options).format("jdbc").load()

    val rdd1 = studentInfo.map(x => (x.getString(0), x.getInt(1)))
    val rdd2 = studentScore.map(x => (x.getString(0), x.getInt(1)))

    val studentRDD = rdd1.join(rdd2).filter(x => x._2._2 > 0)

    studentRDD.foreach(println(_))

    val goodStrudenRow = studentRDD.map(x => Row(x._1, x._2._1, x._2._2))
    //Pairdd转化为df

    //fileds
    val fileds = List(StructField("name", StringType, true),
      StructField("age", IntegerType, true),
      StructField("score", IntegerType, true))

    val df = sqlContext.createDataFrame(goodStrudenRow, StructType(fileds))
    val rowArray = df.collect()
    for (row <- rowArray) {
      println(row)
    }
//    import scala.collection.JavaConversions._
//
//    val lines=  Files.readAllLines(new File("C:\\Users\\WUJUN\\Desktop\\2222222222222222\\students.txt").toPath)


  }


}
